libraries/standard/json/src/json.c

```diff
@@ -0,0 +1,1173 @@
+ *
+ * Iterate over the key-value pairs of an object, looking for a matching key.
+ *
+ * @return #JSONSuccess if the queryKey is found and the value output;
+ * #JSONIllegalDocument if the buffer contents are NOT valid JSON;
+ * #JSONMaxDepthExceeded if object and array nesting exceeds a threshold;
+ * #JSONNotFound if the queryKey is NOT found.
+ *
+ * @note Parsing stops upon finding a match.
+ */
+static JSONStatus_t search( char * buf,
+                            size_t max,
+                            const char * queryKey,
+                            size_t queryKeyLength,
+                            char ** outValue,
+                            size_t * outValueLength )
+{
+    JSONStatus_t ret = JSONPartial;
+    size_t i = 0U, key, keyLength, value, valueLength;
+
+    skipSpace( buf, &i, max );
+
+    if( ( i < max ) && ( buf[ i ] == '{' ) )
+    {
+        i++;
+        skipSpace( buf, &i, max );
+
+        while( i < max )
+        {
+            if( nextKeyValuePair( buf, &i, max, &key, &keyLength,
+                                  &value, &valueLength ) != true )
+            {
+                ret = JSONIllegalDocument;
+                break;
+            }
+
+            if( ( queryKeyLength == keyLength ) &&
+                ( strnEq( queryKey, &buf[ key ], keyLength ) == true ) )
+            {
+                ret = JSONSuccess;
+                break;
+            }
+
+            if( skipSpaceAndComma( buf, &i, max ) != true )
+            {
+                break;
+            }
+        }
+    }
+
+    if( ret == JSONSuccess )
+    {
+        /* String values and collections include their surrounding
+         * demarcation.  If the value is a string, strip the quotes. */
+        if( buf[ value ] == '"' )
+        {
+            value++;
+            valueLength -= 2U;
+        }
+
+        *outValue = &buf[ value ];
+        *outValueLength = valueLength;
+    }
+    else if( ret == JSONPartial )
+    {
+        ret = ( buf[ i ] == '}' ) ? JSONNotFound : JSONIllegalDocument;
+    }
+    else
+    {
+        /* MISRA 15.7 */
+    }
+
+    return ret;
+}
+
+/**
+ * See json.h for docs.
+ *
+ * Handle a nested search by iterating over the parts of the queryKey.
+ */
+JSONStatus_t JSON_Search( char * buf,
+                          size_t max,
+                          char * queryKey,
+                          size_t queryKeyLength,
+                          char separator,
+                          char ** outValue,
+                          size_t * outValueLength )
+{
+    JSONStatus_t ret = JSONPartial;
+    size_t i = 0U, start = 0U, keyLength = 0U;
+    char * p = buf;
+    size_t tmp = max;
+
+    assert( ( buf != NULL ) && ( max > 0U ) );
+    assert( ( queryKey != NULL ) && ( queryKeyLength != 0U ) );
+    assert( ( outValue != NULL ) && ( outValueLength != NULL ) );
+    assert( *queryKey != '\0' );
+
+    while( i < queryKeyLength )
```

Comment:
Probably not too interesting for many documents, but I think this loop will not work quite right for queries that are either an empty string or end with a dot. E.g. if my document is '{"": {"": "I feel empty on the outside."}}', then I would expect querying for:

'' -> '{"": "I feel empty on the outside."}'.
'.' -> "I feel empty on the outside."

but the current code would return:

'' -> JSONPartial (this will happen no matter what is in the document!)
'.' -> '{"": "I feel empty on the outside."}' (missing a level of dereferencing)

Reply:
I think the assert would trip for the empty string - clearly I was thinking that was not a valid use case.  It seems to me that empty keys might be like escaped NULs -  they should probably be explicitly disallowed for the sake of security.